### ISSUE TITLE
How to Visualize Attention Mechanism

### ISSUE THREAD

**Issue by pfeducode** [harlanhong/CVPR2022-DaGAN#42]
Is there any code for visualizing the attention mechanism? Or can you recommend a repository  for reference，Thank you

**Comment by harlanhong** [comment]
I wrote it by meself, you can refer  to the following code:
`       img = image.copy()
        mask = attention_map.reshape(64,64)
        mask = (mask-mask.min())/(mask.max()-mask.min())
        # mask= mask/mask.max()
        mask = cv2.resize(mask, (img.shape[1], img.shape[0]))
        img = np.float32(img) / 255
        heatmap = cv2.applyColorMap(np.uint8(255 * mask), cv2.COLORMAP_JET)
        heatmap = np.float32(heatmap) / 255
        cam = heatmap + np.float32(img)
        # cam = heatmap
        cam = cam / np.max(cam)
        img = np.uint8(255 * cam)`

**Comment by pfeducode** [comment]
Thank you for your timely reply, but I don't understand one thing that Attention shape is [1,4096,4096]. I can't reshape it to (64,64). How can I solve this problem？

**Comment by pfeducode** [comment]
This is my modified code. The result is as follows. Why is my figure red?
![1](https://user-images.githubusercontent.com/100210045/194487054-e88f96f5-7976-46a5-bb3d-56cd079b6ae7.png)
 def visualize_attention(self, prediction,attention):
        #rst=[]
        amap=attention.detach().cpu().numpy()

        img=prediction.detach().cpu().numpy().copy()
        img=np.squeeze(img)
        img=img.transpose((1,2,0))

        mask=amap.reshape(1,4096,64,64)
        mask = np.mean(mask, axis=1) #[1,64,64]
        mask = np.squeeze(mask, axis=0) #[64,64]
        mask=(mask-mask.min())/(mask.max()-mask.min())
        #mask=mask/mask.max()
        mask=cv2.resize(mask,(img.shape[1],img.shape[0]))
        img=np.float32(img)/255
        heatmap=cv2.applyColorMap(np.uint8(255*mask),cv2.COLORMAP_JET)
        heatmap=np.float32(heatmap)/255
        cam=heatmap+np.float32(img)
        cam=cam/np.max(cam)
        img=np.uint8(255*cam)
        # rst.append(img)
        
        return img

![2022-10-07 14_45_31 771445](https://user-images.githubusercontent.com/100210045/194487121-af2fa493-bcf8-46a2-86c5-bb9f36d3586f.jpg)

**Comment by pfeducode** [comment]
This is my modified code. The result is as follows, but my result seems different from yours. Is my result correct?
![1](https://user-images.githubusercontent.com/100210045/194518671-28667941-2cd8-4273-b0a8-0aebb86e5efe.png)
    def visualize_attention(self, source_image,attention):

        amap=attention.detach().cpu().numpy()

        img=source_image.detach().cpu().numpy().copy()
        img=np.squeeze(img)
        img=img.transpose((1,2,0))

        mask=amap.reshape(1,4096,64,64)
        mask = np.mean(mask, axis=1) #[1,64,64]
        mask = np.squeeze(mask, axis=0) #[64,64]
        mask=(mask-mask.min())/(mask.max()-mask.min())
        #mask=mask/mask.max()
        mask=cv2.resize(mask,(img.shape[1],img.shape[0]))
        img=np.float32(img)/255
        heatmap=cv2.applyColorMap(np.uint8(255*mask),cv2.COLORMAP_JET)
        heatmap=np.float32(heatmap)/255
        cam=heatmap+np.float32(img)
        cam=cam/np.max(cam)
        img=np.uint8(255*cam)
        return img

if i use opencv-python to save image,the result as follows
![2022-10-07 17_09_20 482669](https://user-images.githubusercontent.com/100210045/194518930-1d439a51-744e-4eb3-b8b4-100a6dba3aba.jpg)

**Comment by harlanhong** [comment]
you should make sure that your input "source_image" ranges between 0 to 255. It seems that you put these codes in the generator.py, and the source_image in the generator.py is normalised to [0,1]. You should check that.

**Comment by pfeducode** [comment]
I have got the correct picture, thank you for your reply and guidance, and thank you again for your excellent work.
